src/hooks/baseHttp.ts

```diff
@@ -0,0 +1,385 @@
+  httpSpan?: BasicChildSpan;
+
+  requestData?: RequestData;
+
+  requestRandomId: string;
+
+  awsRequestId: string;
+  transactionId: string;
+};
+
+export type ParseHttpRequestOptions = {
+  agent?: Agent;
+  _defaultAgent?: Agent;
+  headers?: Record<string, string>;
+  method?: 'GET' | 'POST';
+  protocol?: string;
+  path?: string;
+  port?: number;
+  defaultPort?: number;
+  hostname?: string;
+  host?: string;
+  uri?: { hostname?: string };
+};
+
+export type RequestData = {
+  host?: string;
+  body?: any;
+  headers?: any;
+  path?: string;
+  truncated?: boolean;
+  port?: number;
+  uri?: string;
+  method?: string;
+  protocol?: string;
+  sendTime?: number;
+};
+
+export type httpRequestCreatedParams = {
+  options?: ParseHttpRequestOptions;
+  url?: string;
+};
+
+export class BaseHttp {
+  /**
+   * Starts an HTTP request tracing span
+   * @param {ParseHttpRequestOptions} options Parameters about the new http request that is being triggered
+   * @param {string} url The URL of the new http request that is being triggered
+   * @returns {HttpRequestTracingConfig} The newly created span, and information required for altering the http request
+   */
+  static onRequestCreated({
+    options,
+    url,
+  }: httpRequestCreatedParams): HttpRequestTracingConfig | undefined {
+    // Gather basic info for creating the HTTP span
+    const host = BaseHttp._getHostFromOptionsOrUrl({ options, url });
+    const headers = options.headers || {};
+    const addedHeaders = {};
+    const { awsRequestId } = TracerGlobals.getHandlerInputs().context;
+    const transactionId = getCurrentTransactionId();
+    const requestRandomId = getRandomId();
+
+    const shouldIgnoreReq =
+      BaseHttp.isBlacklisted(host) || !isValidAlias() || GlobalDurationTimer.isTimePassed();
+    if (shouldIgnoreReq) {
+      return undefined;
+    }
+
+    logger.debug('Starting hook', { host, url, headers });
+
+    const returnData: HttpRequestTracingConfig = {
+      addedHeaders,
+      headers,
+      requestRandomId,
+      awsRequestId,
+      transactionId,
+    };
+
+    const isRequestToAwsService = isAwsService(host);
+    if (isRequestToAwsService && !isKeepHeadersOn()) {
+      const { awsXAmznTraceId } = getAWSEnvironment();
+      addedHeaders['X-Amzn-Trace-Id'] = getPatchedTraceId(awsXAmznTraceId);
+    }
+
+    if (shouldPropagateW3C()) {
+      safeExecute(() => {
+        Object.assign(addedHeaders, getW3CTracerPropagatorAdditionalHeaders(headers));
+      })();
+    }
+
+    if (addedHeaders) {
+      Object.assign(headers, addedHeaders);
+      options.headers = headers;
+    }
+
+    const requestData = BaseHttp.parseHttpRequestOptions(options, url);
+    returnData.requestData = requestData;
+
+    if (isTimeoutTimerEnabled()) {
+      const httpSpan = getHttpSpan(transactionId, awsRequestId, requestRandomId, requestData);
+      SpansContainer.addSpan(httpSpan);
+      returnData.httpSpan = httpSpan;
+    }
+    return returnData;
```

Comment:
Critical code segment!
This is in essence the same code as the start hook in the http lib instrumentation, just made it general to match any http library.